File: primes.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

func sieve(read_in chan int, finish chan bool) {
	defer func() {
		finish <- true
	}()

	divisor := <-read_in
	fmt.Printf("prime %d\n", divisor)

	fork := false
	to_next_sieve := make(chan int)
	for {
		number, open := <-read_in
		if !open {
			close(to_next_sieve)
			return
		}

		if number%divisor == 0 {
			continue
		}

		if !fork {
			fork = true
			child_finish := make(chan bool)
			defer func() {
				_ = <-child_finish
			}()
			go sieve(to_next_sieve, child_finish)
		}
		to_next_sieve <- number
	}
}

func main() {
	if len(os.Args) != 2 {
		panic("primes.go expects 1 argument")
	}
	upper, _ := strconv.Atoi(os.Args[1])

	to_sieve := make(chan int)
	finish := make(chan bool)

	go sieve(to_sieve, finish)
	for number := 2; number < upper; number++ {
		to_sieve <- number
	}
	close(to_sieve)
	_ = <-finish
}

File: primes-c/primes.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/wait.h>
#include <unistd.h>

void sieve(int port_in) {
    int pid = -1;
    int fd[2];
    pipe(fd);

    int divisor;
    read(port_in, &divisor, sizeof(divisor));
    printf("prime %d\n", divisor);
    fflush(stdout);

    int number;
    while (read(port_in, &number, sizeof(number)) != 0) {
        if (number % divisor == 0) {
            continue;
        }

        if (pid == -1) {
            // execute only once to create child
            pid = fork();
            if (pid == 0) {
                // child process
                close(fd[1]);
                sieve(fd[0]);
                close(fd[0]);
                exit(0);
            } else {
                // parent process
                close(fd[0]);
            }
        }
        // send number to next sieve
        write(fd[1], &number, sizeof(number));
    }

    close(fd[1]);
    wait(&pid);
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        printf("error: primes expects 1 argument\n");
        exit(1);
    }

    int fd[2];
    pipe(fd);

    int pid = fork();
    if (pid == 0) {
        // child process
        close(fd[1]);
        sieve(fd[0]);
        close(fd[0]);
        exit(0);
    } else {
        // parent process
        close(fd[0]);
        for (int number = 2; number < atoi(argv[1]); number++) {
            write(fd[1], &number, sizeof(number));
        }
        close(fd[1]);
        wait(&pid);
    }

    return 0;
}
